File: 02_regressao_cqr/backtest_cqr.jl
"""
Backtest de intervalos de previsão para retornos diários do Ibovespa.

Compara 4 métodos para construir intervalos 90% sobre retornos t+1:
  (1) Paramétrico: μ ± 1.645·σ rolling de 252 dias.
  (2) Bootstrap dos resíduos sobre regressão linear (1000 reps, percentis 5/95).
  (3) Conformalized Quantile Regression (CQR) com LightGBM como base.
  (4) CV+ (variante jackknife+ com K=10 folds) com LightGBM.

Dataset: IBOV diário 2004-01-05 a 2024-12-30 (^BVSP via yfinance).
Split:
  - Treino: 2004-2018
  - Calibração: 2019 (para CQR; CV+ usa treino+calib internamente)
  - Teste: 2020-2024 (atravessa COVID, eleições e regime alta de juros)

Output: tabela de métricas (cobertura, largura) e cobertura condicional por decil
de volatilidade realizada, figura de bandas em fev-abr 2020.
"""

using CSV, DataFrames, Dates, Random, Statistics, Printf
using LightGBM
using Plots

Random.seed!(42)

const ALPHA = 0.10           # alvo de cobertura: 90%
const HALF  = ALPHA / 2       # 5% / 95%
const Z90   = 1.6448536269514722

# ----------------------------------------------------------------------
# 1. Carregar dados e construir features
# ----------------------------------------------------------------------
df = CSV.read("ibov_returns.csv", DataFrame)
rename!(df, :Date => :date)
sort!(df, :date)

# Features: lags de retorno + volatilidade realizada
function rolling_std(x::Vector{Float64}, w::Int)
    n = length(x)
    out = fill(NaN, n)
    @inbounds for i in w:n
        out[i] = std(view(x, (i-w+1):i))
    end
    return out
end

r = df.ret
df.ret_lag1   = [NaN; r[1:end-1]]
df.ret_lag5   = [fill(NaN, 5); r[1:end-5]]
df.ret_lag22  = [fill(NaN, 22); r[1:end-22]]
df.rv_5       = [NaN; rolling_std(r, 5)[1:end-1]]
df.rv_22      = [NaN; rolling_std(r, 22)[1:end-1]]
df.absret_lag1 = abs.(df.ret_lag1)
df.absret_lag5 = abs.(df.ret_lag5)

dropmissing!(df)
df = df[.!any.(eachrow(isnan.(df[:, Not(:date)]))), :]

println("After feature build: $(nrow(df)) rows from $(df.date[1]) to $(df.date[end])")

const FEATURES = [:ret_lag1, :ret_lag5, :ret_lag22, :rv_5, :rv_22, :absret_lag1, :absret_lag5]

# Splits temporais
train = df[df.date .< Date(2019,1,1), :]
cal   = df[(df.date .>= Date(2019,1,1)) .& (df.date .< Date(2020,1,1)), :]
test  = df[df.date .>= Date(2020,1,1), :]
println("Train: $(nrow(train)) | Cal: $(nrow(cal)) | Test: $(nrow(test))")

X_train = Matrix(train[:, FEATURES])
y_train = train.ret
X_cal   = Matrix(cal[:, FEATURES])
y_cal   = cal.ret
X_test  = Matrix(test[:, FEATURES])
y_test  = test.ret
X_full  = vcat(X_train, X_cal)
y_full  = vcat(y_train, y_cal)
rv_test = test.rv_22

# ----------------------------------------------------------------------
# Método 1: Paramétrico (rolling normal 252d)
# ----------------------------------------------------------------------
function method_parametric(df_all::DataFrame, test_idx::Vector{Date})
    r = df_all.ret
    n = length(r)
    μ = fill(NaN, n); σ = fill(NaN, n)
    @inbounds for i in 252:n
        win = view(r, (i-251):i)
        μ[i] = mean(win); σ[i] = std(win)
    end
    full_idx = df_all.date
    lo = fill(NaN, length(test_idx)); hi = fill(NaN, length(test_idx))
    for (k, d) in enumerate(test_idx)
        j = searchsortedfirst(full_idx, d)
        if j <= n && full_idx[j] == d
            lo[k] = μ[j] - Z90*σ[j]
            hi[k] = μ[j] + Z90*σ[j]
        end
    end
    return lo, hi
end

# ----------------------------------------------------------------------
# Método 2: Bootstrap dos resíduos sobre regressão linear
# ----------------------------------------------------------------------
function ols_fit(X::Matrix{Float64}, y::Vector{Float64})
    Xi = hcat(ones(size(X,1)), X)
    β = Xi \ y
    return β
end
ols_predict(β::Vector{Float64}, X::Matrix{Float64}) = hcat(ones(size(X,1)), X) * β

function method_bootstrap(B::Int=1000)
    β = ols_fit(X_train, y_train)
    resid = y_train .- ols_predict(β, X_train)
    pred_test = ols_predict(β, X_test)
    rng = MersenneTwister(42)
    n_test = length(y_test)
    samples = Matrix{Float64}(undef, B, n_test)
    for b in 1:B
        bs = resid[rand(rng, 1:length(resid), n_test)]
        samples[b, :] = pred_test .+ bs
    end
    lo = [quantile(view(samples, :, j), HALF)     for j in 1:n_test]
    hi = [quantile(view(samples, :, j), 1 - HALF) for j in 1:n_test]
    return lo, hi
end

# ----------------------------------------------------------------------
# Método 3: CQR (Conformalized Quantile Regression) com LightGBM
# ----------------------------------------------------------------------
function fit_lgb_quantile(X, y, q)
    m = LGBMRegression(
        objective = "quantile", alpha = q,
        num_leaves = 31, learning_rate = 0.05,
        num_iterations = 300, min_data_in_leaf = 20,
        verbosity = -1,
    )
    LightGBM.fit!(m, X, Float64.(y))
    return m
end

function method_cqr()
    q_lo_model = fit_lgb_quantile(X_train, y_train, HALF)
    q_hi_model = fit_lgb_quantile(X_train, y_train, 1 - HALF)

    q_lo_cal = vec(LightGBM.predict(q_lo_model, X_cal))
    q_hi_cal = vec(LightGBM.predict(q_hi_model, X_cal))
    E = max.(q_lo_cal .- y_cal, y_cal .- q_hi_cal)

    n = length(y_cal)
    level = min(ceil((1 - ALPHA)*(n + 1))/n, 1.0)
    Q = quantile(E, level)

    q_lo_test = vec(LightGBM.predict(q_lo_model, X_test))
    q_hi_test = vec(LightGBM.predict(q_hi_model, X_test))
    return q_lo_test .- Q, q_hi_test .+ Q
end

# ----------------------------------------------------------------------
# Método 4: CV+ (K=10) com LightGBM regressão padrão
# ----------------------------------------------------------------------
function kfold_indices(n::Int, K::Int, rng::AbstractRNG)
    perm = randperm(rng, n)
    fold_size = div(n, K)
    folds = Vector{Vector{Int}}(undef, K)
    for k in 1:K
        s = (k - 1)*fold_size + 1
        e = k == K ? n : k*fold_size
        folds[k] = perm[s:e]
    end
    return folds
end

function method_cvplus(K::Int=10)
    rng = MersenneTwister(42)
    n_train = size(X_full, 1)
    n_test  = size(X_test, 1)

    folds = kfold_indices(n_train, K, rng)
    R        = zeros(n_train)
    fold_id  = zeros(Int, n_train)
    pred_test_per_fold = zeros(K, n_test)

    for (k, val_idx) in enumerate(folds)
        tr_idx = setdiff(1:n_train, val_idx)
        m = LGBMRegression(
            objective = "regression",
            num_leaves = 31, learning_rate = 0.05,
            num_iterations = 300, min_data_in_leaf = 20,
            verbosity = -1,
        )
        LightGBM.fit!(m, X_full[tr_idx, :], Float64.(y_full[tr_idx]))
        ŷ_val = vec(LightGBM.predict(m, X_full[val_idx, :]))
        R[val_idx] .= abs.(y_full[val_idx] .- ŷ_val)
        fold_id[val_idx] .= k
        pred_test_per_fold[k, :] = vec(LightGBM.predict(m, X_test))
    end

    # Para cada x_test, distribuição empírica de {f̂^{-k(i)}(x_test) ± R_i}
    pred_at_x_for_each_i = pred_test_per_fold[fold_id, :]   # (n_train, n_test)
    L = pred_at_x_for_each_i .- R
    H = pred_at_x_for_each_i .+ R

    lo = [quantile(view(L, :, j), HALF)     for j in 1:n_test]
    hi = [quantile(view(H, :, j), 1 - HALF) for j in 1:n_test]
    return lo, hi
end

# ----------------------------------------------------------------------
# Executar e medir
# ----------------------------------------------------------------------
println("Fitting Parametric ..."); lo1, hi1 = method_parametric(df, test.date)
println("Fitting Bootstrap ...");  lo2, hi2 = method_bootstrap()
println("Fitting CQR ...");        lo3, hi3 = method_cqr()
println("Fitting CV+ ...");        lo4, hi4 = method_cvplus(10)

methods = [
    ("Paramétrico (rolling 252d)", lo1, hi1),
    ("Bootstrap (LR + 1000 reps)", lo2, hi2),
    ("CQR (LightGBM-quantile)",    lo3, hi3),
    ("CV+ (K=10, LightGBM)",       lo4, hi4),
]

function metrics(lo, hi, y, rv)
    mask = .!(isnan.(lo) .| isnan.(hi))
    y_, lo_, hi_, rv_ = y[mask], lo[mask], hi[mask], rv[mask]
    cov   = mean((y_ .>= lo_) .& (y_ .<= hi_))
    width = mean(hi_ .- lo_)
    # Cobertura condicional por decil de RV
    cuts = quantile(rv_, range(0, 1; length=11))
    decile = clamp.(searchsortedfirst.(Ref(cuts[2:end]), rv_), 1, 10)
    cond = [mean(((y_ .>= lo_) .& (y_ .<= hi_))[decile .== d]) for d in 1:10]
    return cov, width, cond, sum(mask)
end

results = Dict{String, NamedTuple}()
for (name, lo, hi) in methods
    cov, w, cond, n_eff = metrics(lo, hi, y_test, rv_test)
    results[name] = (coverage=cov, width=w, cond=cond, n=n_eff)
    @printf("%-30s cov=%.3f  width=%.4f  n=%d\n", name, cov, w, n_eff)
end

# Salvar tabelas
summary_df = DataFrame(
    method   = [name for (name, _, _) in methods],
    coverage = [results[name].coverage for (name, _, _) in methods],
    width    = [results[name].width    for (name, _, _) in methods],
    n        = [results[name].n        for (name, _, _) in methods],
)
CSV.write("results_summary.csv", summary_df)

cond_df = DataFrame(decile=["D$i" for i in 1:10])
for (name, _, _) in methods
    cond_df[!, name] = results[name].cond
end
CSV.write("results_conditional.csv", cond_df)

println("\nConditional coverage by RV decile (D1=lowest vol, D10=highest):")
show(stdout, "text/plain", cond_df); println()
println("\nSaved results_summary.csv and results_conditional.csv")

# ----------------------------------------------------------------------
# Figura: banda paramétrica vs banda CQR em fev-abr 2020
# ----------------------------------------------------------------------
win_start = Date(2020,2,3)
win_end   = Date(2020,4,30)
mask_win  = (test.date .>= win_start) .& (test.date .<= win_end)
dates_win = test.date[mask_win]
y_win     = y_test[mask_win]
lo_par_w  = lo1[mask_win]; hi_par_w = hi1[mask_win]
lo_cqr_w  = lo3[mask_win]; hi_cqr_w = hi3[mask_win]

p = plot(size=(1000, 520), dpi=150, legend=:topright, grid=true, gridalpha=0.25,
         ylabel="Retorno diário (%)", xlabel="", framestyle=:box)
plot!(p, dates_win, 100 .* hi_par_w, fillrange=100 .* lo_par_w,
      fillcolor="#888888", fillalpha=0.30, linecolor="#555555", linewidth=0.9,
      label="Paramétrico (rolling 252d)")
plot!(p, dates_win, 100 .* lo_par_w, linecolor="#555555", linewidth=0.9, label="")
plot!(p, dates_win, 100 .* hi_cqr_w, fillrange=100 .* lo_cqr_w,
      fillcolor="#FF6719", fillalpha=0.25, linecolor="#FF6719", linewidth=1.2,
      label="CQR (LightGBM-quantile)")
plot!(p, dates_win, 100 .* lo_cqr_w, linecolor="#FF6719", linewidth=1.2, label="")
plot!(p, dates_win, 100 .* y_win, linecolor="#0a1420", linewidth=1.3,
      marker=:circle, markersize=2.5, label="Retorno realizado IBOV")
hline!(p, [0], linecolor="#999", linewidth=0.6, label="")
target = Date(2020,3,18)
vline!(p, [target], linecolor="#c0392b", linewidth=0.8, linestyle=:dash, label="")

savefig(p, "band_march_2020.png")
println("Saved band_march_2020.png")

# Valores específicos de 18-mar-2020 (para uso no artigo)
ti = test.date
if target in ti
    j = findfirst(==(target), ti)
    println("\n=== 18/03/2020 ===")
    @printf("  Retorno realizado:  %+.2f%%\n", 100*y_test[j])
    @printf("  Paramétrico:        [%+.2f%%, %+.2f%%]  largura=%.2f%%\n",
            100*lo1[j], 100*hi1[j], 100*(hi1[j]-lo1[j]))
    @printf("  Bootstrap residual: [%+.2f%%, %+.2f%%]  largura=%.2f%%\n",
            100*lo2[j], 100*hi2[j], 100*(hi2[j]-lo2[j]))
    @printf("  CQR:                [%+.2f%%, %+.2f%%]  largura=%.2f%%\n",
            100*lo3[j], 100*hi3[j], 100*(hi3[j]-lo3[j]))
    @printf("  CV+:                [%+.2f%%, %+.2f%%]  largura=%.2f%%\n",
            100*lo4[j], 100*hi4[j], 100*(hi4[j]-lo4[j]))
end

File: 02_regressao_cqr/backtest_cqr.py
"""
Backtest de intervalos de previsão para retornos diários do Ibovespa.

Compara métodos para construir intervalos 90% sobre retornos t+1:
  (1) Paramétrico: μ ± 1.645·σ rolling de 252 dias.
  (2) Bootstrap em blocos (Künsch, 1989), L=22 dias, 1000 reamostragens.
  (3) CQR com LightGBM-quantile como base (Romano, Patterson, Candès, 2019;
      LightGBM de Ke et al., 2017).
  (4) CQR com regressão quantílica linear como base (Koenker e Bassett, 1978).
  (5) CV+ (variante jackknife+ com K=10 folds) com LightGBM como regressor.

Dataset: IBOV diário 2004-01-05 a 2025-12-30 (^BVSP via yfinance).
Splits:
  - Treino: 2004-2013 (~2449 dias)
  - Calibração: 2014 (~248 dias, usada pelas CQR; CV+ une treino+cal)
  - Teste: 2015-2025 (~2729 dias). Cobre crise fiscal 2015-16,
                     Joesley 2017, eleição 2018, COVID, ciclo de
                     juros e fiscal 2022-2025.

Saída:
  - results_summary.csv: cobertura global e largura média por método.
  - results_conditional.csv: cobertura por decil de volatilidade realizada.
  - band_march_2020.png: bandas paramétrica e CQR em fev-abr/2020.

Reprodutibilidade: seed 42 em todo o pipeline.
"""
import os
import numpy as np
import pandas as pd
import lightgbm as lgb
import statsmodels.api as sm
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.dates import DateFormatter

np.random.seed(42)

ALPHA = 0.10
HALF = ALPHA / 2
Z90 = 1.6448536269514722

HERE = os.path.dirname(os.path.abspath(__file__))


# ----------------------------------------------------------------------
# 1. Carregar dados e construir features
# ----------------------------------------------------------------------
df = pd.read_csv(os.path.join(HERE, "ibov_returns.csv"), parse_dates=["Date"])
df = df.rename(columns={"Date": "date"}).sort_values("date").reset_index(drop=True)
r = df["ret"].values
n = len(r)


def shift_lag(x, k):
    out = np.full(n, np.nan)
    out[k:] = x[:-k]
    return out


def rolling_std(x, w):
    out = np.full(n, np.nan)
    for i in range(w, n + 1):
        out[i - 1] = np.std(x[i - w:i], ddof=1)
    return out


df["ret_lag1"]   = shift_lag(r, 1)
df["ret_lag5"]   = shift_lag(r, 5)
df["ret_lag22"]  = shift_lag(r, 22)
df["rv_5"]       = shift_lag(rolling_std(r, 5), 1)
df["rv_22"]      = shift_lag(rolling_std(r, 22), 1)
df["absret_lag1"] = np.abs(df["ret_lag1"])
df["absret_lag5"] = np.abs(df["ret_lag5"])
df = df.dropna().reset_index(drop=True)
print(f"After feature build: {len(df)} rows from {df.date.iloc[0].date()} to {df.date.iloc[-1].date()}")

FEATURES = ["ret_lag1", "ret_lag5", "ret_lag22", "rv_5", "rv_22", "absret_lag1", "absret_lag5"]

train = df[df.date < pd.Timestamp(2014, 1, 1)].copy()
cal   = df[(df.date >= pd.Timestamp(2014, 1, 1)) & (df.date < pd.Timestamp(2015, 1, 1))].copy()
test  = df[df.date >= pd.Timestamp(2015, 1, 1)].copy()
print(f"Train: {len(train)} | Cal: {len(cal)} | Test: {len(test)}")

X_train = train[FEATURES].values
y_train = train["ret"].values
X_cal   = cal[FEATURES].values
y_cal   = cal["ret"].values
X_test  = test[FEATURES].values
y_test  = test["ret"].values
X_full  = np.vstack([X_train, X_cal])
y_full  = np.concatenate([y_train, y_cal])
rv_test = test["rv_22"].values
test_dates = test.date.values


# ----------------------------------------------------------------------
# Método 1: Paramétrico (rolling normal 252d)
# ----------------------------------------------------------------------
def method_parametric():
    r_all = df["ret"].values
    n_all = len(r_all)
    mu = np.full(n_all, np.nan)
    sig = np.full(n_all, np.nan)
    for i in range(252, n_all + 1):
        win = r_all[i - 252:i]
        mu[i - 1] = np.mean(win)
        sig[i - 1] = np.std(win, ddof=1)
    full_dates = df.date.values
    lo = np.full(len(test_dates), np.nan)
    hi = np.full(len(test_dates), np.nan)
    full_dates_idx = pd.Series(np.arange(n_all), index=full_dates)
    for k, d in enumerate(test_dates):
        if d in full_dates_idx.index:
            j = full_dates_idx[d]
            lo[k] = mu[j] - Z90 * sig[j]
            hi[k] = mu[j] + Z90 * sig[j]
    return lo, hi


# ----------------------------------------------------------------------
# Método 2: Bootstrap em blocos (Künsch, 1989) sobre regressão linear
# ----------------------------------------------------------------------
def ols_fit(X, y):
    Xi = np.hstack([np.ones((X.shape[0], 1)), X])
    return np.linalg.lstsq(Xi, y, rcond=None)[0]


def ols_predict(beta, X):
    return np.hstack([np.ones((X.shape[0], 1)), X]) @ beta


def method_block_bootstrap(B=1000, L=22):
    beta = ols_fit(X_train, y_train)
    resid = y_train - ols_predict(beta, X_train)
    pred_test = ols_predict(beta, X_test)
    rng = np.random.RandomState(42)
    n_test = len(y_test)
    n_resid = len(resid)
    n_blocks = int(np.ceil(n_test / L))
    samples = np.zeros((B, n_test))
    for b in range(B):
        starts = rng.randint(0, n_resid - L + 1, size=n_blocks)
        bs = np.concatenate([resid[s:s + L] for s in starts])[:n_test]
        samples[b, :] = pred_test + bs
    lo = np.quantile(samples, HALF, axis=0)
    hi = np.quantile(samples, 1 - HALF, axis=0)
    return lo, hi


# ----------------------------------------------------------------------
# Métodos 3 e 4: CQR com base LightGBM e CQR com base linear (Koenker-Bassett)
# ----------------------------------------------------------------------
def fit_lgb_quantile(X, y, q):
    m = lgb.LGBMRegressor(
        objective="quantile", alpha=q,
        num_leaves=31, learning_rate=0.05,
        n_estimators=300, min_data_in_leaf=20,
        verbose=-1, random_state=42,
    )
    m.fit(X, y)
    return m


def _conformal_adjust(q_lo_cal, q_hi_cal, y_cal_):
    """Calcula Q̂ conformal a partir dos scores E_i = max(qlo - y, y - qhi)."""
    E = np.maximum(q_lo_cal - y_cal_, y_cal_ - q_hi_cal)
    n_cal = len(y_cal_)
    level = min(np.ceil((1 - ALPHA) * (n_cal + 1)) / n_cal, 1.0)
    return float(np.quantile(E, level, method="higher"))


def method_cqr_lgb():
    q_lo_model = fit_lgb_quantile(X_train, y_train, HALF)
    q_hi_model = fit_lgb_quantile(X_train, y_train, 1 - HALF)
    q_lo_cal = q_lo_model.predict(X_cal)
    q_hi_cal = q_hi_model.predict(X_cal)
    Q = _conformal_adjust(q_lo_cal, q_hi_cal, y_cal)
    q_lo_test = q_lo_model.predict(X_test)
    q_hi_test = q_hi_model.predict(X_test)
    return q_lo_test - Q, q_hi_test + Q, Q


def method_cqr_linear():
    """CQR usando regressão quantílica linear (Koenker e Bassett, 1978) como base."""
    X_train_c = sm.add_constant(X_train)
    X_cal_c   = sm.add_constant(X_cal)
    X_test_c  = sm.add_constant(X_test)
    beta_lo = sm.QuantReg(y_train, X_train_c).fit(q=HALF,     max_iter=5000).params
    beta_hi = sm.QuantReg(y_train, X_train_c).fit(q=1 - HALF, max_iter=5000).params
    q_lo_cal = X_cal_c @ beta_lo
    q_hi_cal = X_cal_c @ beta_hi
    Q = _conformal_adjust(q_lo_cal, q_hi_cal, y_cal)
    q_lo_test = X_test_c @ beta_lo
    q_hi_test = X_test_c @ beta_hi
    return q_lo_test - Q, q_hi_test + Q, Q


# ----------------------------------------------------------------------
# Método 5: CV+ (K=10) com LightGBM (usado pelo artigo 02b)
# ----------------------------------------------------------------------
def method_cvplus(K=10):
    rng = np.random.RandomState(42)
    n_train = X_full.shape[0]
    n_test  = X_test.shape[0]
    perm = rng.permutation(n_train)
    fold_size = n_train // K
    folds = []
    for k in range(K):
        s = k * fold_size
        e = n_train if k == K - 1 else (k + 1) * fold_size
        folds.append(perm[s:e])
    R = np.zeros(n_train)
    fold_id = np.zeros(n_train, dtype=int)
    pred_test_per_fold = np.zeros((K, n_test))
    for k, val_idx in enumerate(folds):
        tr_idx = np.setdiff1d(np.arange(n_train), val_idx)
        m = lgb.LGBMRegressor(
            objective="regression",
            num_leaves=31, learning_rate=0.05,
            n_estimators=300, min_data_in_leaf=20,
            verbose=-1, random_state=42,
        )
        m.fit(X_full[tr_idx], y_full[tr_idx])
        R[val_idx] = np.abs(y_full[val_idx] - m.predict(X_full[val_idx]))
        fold_id[val_idx] = k
        pred_test_per_fold[k] = m.predict(X_test)
    pred_at_x = pred_test_per_fold[fold_id, :]
    L_mat = pred_at_x - R[:, None]
    H_mat = pred_at_x + R[:, None]
    lo = np.quantile(L_mat, HALF, axis=0)
    hi = np.quantile(H_mat, 1 - HALF, axis=0)
    return lo, hi


# ----------------------------------------------------------------------
# Executar
# ----------------------------------------------------------------------
print("Fitting Parametric ...");        lo1, hi1 = method_parametric()
print("Fitting Block bootstrap ...");   lo2, hi2 = method_block_bootstrap()
print("Fitting CQR-LightGBM ...");      lo3, hi3, Q_lgb    = method_cqr_lgb()
print("Fitting CQR-Linear (KB) ...");   lo4, hi4, Q_lin    = method_cqr_linear()
print("Fitting CV+ ...");               lo5, hi5 = method_cvplus()

methods = [
    ("Paramétrico (rolling 252d)",            lo1, hi1),
    ("Bootstrap blocos (LR, L=22, 1000 reps)", lo2, hi2),
    ("CQR-LightGBM (quantile)",               lo3, hi3),
    ("CQR-Linear (Koenker-Bassett)",          lo4, hi4),
    ("CV+ (K=10, LightGBM)",                  lo5, hi5),
]


def metrics(lo, hi, y, rv):
    mask = ~(np.isnan(lo) | np.isnan(hi))
    y_, lo_, hi_, rv_ = y[mask], lo[mask], hi[mask], rv[mask]
    cov = float(np.mean((y_ >= lo_) & (y_ <= hi_)))
    width = float(np.mean(hi_ - lo_))
    cuts = np.quantile(rv_, np.linspace(0, 1, 11))
    decile = np.clip(np.searchsorted(cuts[1:], rv_, side="left") + 1, 1, 10)
    cond = [float(np.mean((y_[decile == d] >= lo_[decile == d]) & (y_[decile == d] <= hi_[decile == d])))
            for d in range(1, 11)]
    return cov, width, cond, int(mask.sum())


results = {}
for name, lo, hi in methods:
    cov, w, cond, n_eff = metrics(lo, hi, y_test, rv_test)
    results[name] = dict(coverage=cov, width=w, cond=cond, n=n_eff)
    print(f"{name:42s} cov={cov:.4f}  width={w:.4f}  n={n_eff}")

# CSVs canônicos
summary_df = pd.DataFrame({
    "method":   [m[0] for m in methods],
    "coverage": [results[m[0]]["coverage"] for m in methods],
    "width":    [results[m[0]]["width"]    for m in methods],
    "n":        [results[m[0]]["n"]        for m in methods],
})
summary_df.to_csv(os.path.join(HERE, "results_summary.csv"), index=False)

cond_df = pd.DataFrame({"decile": [f"D{i}" for i in range(1, 11)]})
for name, _, _ in methods:
    cond_df[name] = results[name]["cond"]
cond_df.to_csv(os.path.join(HERE, "results_conditional.csv"), index=False)

print("\nresults_summary.csv:")
print(summary_df.to_string(index=False))
print("\nresults_conditional.csv:")
print(cond_df.to_string(index=False))

# Figura: banda paramétrica vs CQR-LightGBM em fev-abr 2020
win_start = pd.Timestamp(2020, 2, 3)
win_end   = pd.Timestamp(2020, 4, 30)
mask_win = (test.date >= win_start) & (test.date <= win_end)
dates_win = pd.to_datetime(test.date[mask_win])
y_win = y_test[mask_win.values]
lo_par_w = lo1[mask_win.values]; hi_par_w = hi1[mask_win.values]
lo_cqr_w = lo3[mask_win.values]; hi_cqr_w = hi3[mask_win.values]

fig, ax = plt.subplots(figsize=(10, 5.2), dpi=150)
ax.fill_between(dates_win, 100 * lo_par_w, 100 * hi_par_w, color="#888888", alpha=0.30,
                edgecolor="#555555", linewidth=0.9, label="Paramétrico (rolling 252d)")
ax.fill_between(dates_win, 100 * lo_cqr_w, 100 * hi_cqr_w, color="#FF6719", alpha=0.25,
                edgecolor="#FF6719", linewidth=1.2, label="CQR-LightGBM")
ax.plot(dates_win, 100 * y_win, color="#0a1420", linewidth=1.3,
        marker="o", markersize=2.8, label="Retorno realizado IBOV")
ax.axhline(0, color="#999", linewidth=0.6)
ax.axvline(pd.Timestamp(2020, 3, 18), color="#c0392b", linewidth=0.8, linestyle="--")
ax.set_ylabel("Retorno diário (%)")
ax.legend(loc="upper right")
ax.grid(True, alpha=0.25)
ax.xaxis.set_major_formatter(DateFormatter("%d-%b"))
plt.tight_layout()
plt.savefig(os.path.join(HERE, "band_march_2020.png"))
plt.close(fig)
print(f"\nSaved band_march_2020.png")

# Detalhe em 18-mar-2020
target = pd.Timestamp(2020, 3, 18)
idx_arr = np.where(test_dates == np.datetime64(target.date()))[0]
if len(idx_arr) > 0:
    j = idx_arr[0]
    print(f"\n=== 18/03/2020 ===")
    print(f"  Retorno realizado:    {100*y_test[j]:+.2f}%")
    print(f"  Paramétrico:          [{100*lo1[j]:+.2f}%, {100*hi1[j]:+.2f}%]  largura={100*(hi1[j]-lo1[j]):.2f}%")
    print(f"  Bootstrap blocos:     [{100*lo2[j]:+.2f}%, {100*hi2[j]:+.2f}%]  largura={100*(hi2[j]-lo2[j]):.2f}%")
    print(f"  CQR-LightGBM:         [{100*lo3[j]:+.2f}%, {100*hi3[j]:+.2f}%]  largura={100*(hi3[j]-lo3[j]):.2f}%")
    print(f"  CQR-Linear (KB):      [{100*lo4[j]:+.2f}%, {100*hi4[j]:+.2f}%]  largura={100*(hi4[j]-lo4[j]):.2f}%")
    print(f"  CV+:                  [{100*lo5[j]:+.2f}%, {100*hi5[j]:+.2f}%]  largura={100*(hi5[j]-lo5[j]):.2f}%")

print(f"\nQ̂ conformal: CQR-LightGBM = {Q_lgb:+.4f}, CQR-Linear = {Q_lin:+.4f}")
